Code/Tools/FBuild/FBuildCore/Helpers/CompilationDatabase.cpp

```diff
@@ -0,0 +1,215 @@
+                break;
+            }
+            case Node::LIBRARY_NODE:
+            {
+                HandleObjectListNode( node->CastTo< LibraryNode >() );
+                break;
+            }
+            default: break;
+        }
+    }
+}
+
+// HandleObjectListNode
+//------------------------------------------------------------------------------
+void CompilationDatabase::HandleObjectListNode( ObjectListNode * node )
+{
+    ObjectListContext ctx;
+    ctx.m_DB = this;
+    ctx.m_ObjectListNode = node;
+
+    const CompilerNode * compiler = node->GetCompiler();
+    const bool isMSVC = ( compiler->GetCompilerFamily() == CompilerNode::MSVC );
+
+    ctx.m_CompilerEscaped = compiler->GetExecutable();
+    ctx.m_CompilerEscaped.JSONEscape();
+
+    // Prepare arguments: tokenize, remove problematic arguments, remove extra quoting and escape.
+    node->GetCompilerOptions().Tokenize( ctx.m_ArgumentsEscaped );
+    for ( size_t i = 0; i < ctx.m_ArgumentsEscaped.GetSize(); ++i )
+    {
+        AString & argument = ctx.m_ArgumentsEscaped[ i ];
+        if ( isMSVC && ObjectNode::IsStartOfCompilerArg_MSVC( argument, "analyze" ) )
+        {
+            // Clang doesn't recognize /analyze as a compiler option, treats it an input file and
+            // later complains that /Fo option can't be used with multiple input files.
+            // To avoid these problems we strip this option.
+            ctx.m_ArgumentsEscaped.EraseIndex( i-- );
+            continue;
+        }
+        argument.Unquote();
+        argument.JSONEscape();
+    }
+
+    node->EnumerateInputFiles( &CompilationDatabase::HandleInputFile, &ctx );
+}
+
+// HandleInputFile
+//------------------------------------------------------------------------------
+/*static*/ void CompilationDatabase::HandleInputFile( const AString & inputFile, const AString & baseDir, void * userData )
+{
+    ObjectListContext * ctx = static_cast< ObjectListContext * >( userData );
+    ctx->m_DB->HandleInputFile( inputFile, baseDir, ctx );
+}
+
+// HandleInputFile
+//------------------------------------------------------------------------------
+void CompilationDatabase::HandleInputFile( const AString & inputFile, const AString & baseDir, ObjectListContext * ctx )
+{
+    AStackString<> inputFileEscaped;
+    inputFileEscaped = inputFile;
+    inputFileEscaped.JSONEscape();
+
+    AStackString<> outputFileEscaped;
+    ctx->m_ObjectListNode->GetObjectFileName( inputFile, baseDir, outputFileEscaped );
+    outputFileEscaped.JSONEscape();
+
+    m_Output += "  {\n    \"directory\": \"";
+    m_Output += m_DirectoryEscaped;
+    m_Output += "\",\n    \"file\": \"";
+    m_Output += inputFileEscaped;
+    m_Output += "\",\n    \"output\": \"";
+    m_Output += outputFileEscaped;
+    m_Output += "\",\n    \"arguments\": [\"";
+    m_Output += ctx->m_CompilerEscaped;
+    m_Output += "\"";
+    for ( const AString & argument : ctx->m_ArgumentsEscaped )
+    {
```

Comment:
This code seems fragile. Any changes to how args are built for the compiler won't be reflected here. It might be that anything not currently handled here isn't important, but this is sure to get broken if the %1 syntax is expanded in the future.

Reply:
As I wrote in the original post, compilation database doesn't have to have the same exact arguments that are used during the actual build, it has to have some arguments that allow to correctly parse the sources.

So almost all of the argument manipulations that `ObjectNode::BuildArgs` does are not wanted here, arguments specified explicitly in `.CompilerOptions` will be sufficient.

The only exception is expansion of `%N` tokens, but while the logic for it is the same (replace `%N` with some string) the actual code is completely different due to different argument storage (`AString` vs `Args`). So I don't think that this part can be easily refactored to be used from both places.